File: frontend/src/mocks/handlers/loans.ts
import { delay, http, HttpResponse } from 'msw';
import { mockError } from '../helpers';
import type { Loan, LoanBookingRequest, LoanListResponse } from '@/types/loan';

const MOCK_PRICES: Record<string, string> = { BTC: '63500.00', ETH: '1700.00' };

let mockLoans: Loan[] = [];

export function resetMockLoans(): void {
  mockLoans = [];
}

export const loanHandlers = [
  http.get('/api/market-data/prices/:assetType', ({ params }) => {
    const assetType = params.assetType as string;
    const price = MOCK_PRICES[assetType] ?? '0.00';
    return HttpResponse.json({
      asset_type: assetType,
      price_usd: price,
      as_of: new Date().toISOString(),
    });
  }),

  http.post('/api/loans', async ({ request }) => {
    await delay(20);
    const body = (await request.json().catch(() => null)) as LoanBookingRequest | null;
    if (!body?.connection_id || !body.borrower_id || !body.asset_type || !body.quantity) {
      return mockError('validation_error', 'Missing loan booking fields', 422);
    }

    const loanId = `loan-${Date.now()}`;
    const loan: Loan = {
      loan_id: loanId,
      connection_id: body.connection_id,
      agreement_id: 'agreement-001',
      borrower_id: body.borrower_id,
      borrower_name: 'Booked Borrower',
      asset_type: body.asset_type,
      quantity: body.quantity,
      rate_bps: body.rate_bps,
      term_type: body.term_type,
      maturity_date: body.maturity_date,
      day_count_basis: 'actual_360',
      collateral_type: body.collateral_type,
      collateral_quantity: body.collateral_quantity,
      collateral_value_usd: body.collateral_value_usd,
      current_ltv_pct: '30.0',
      ltv_as_of: new Date().toISOString(),
      state: 'pending',
      booked_at: new Date().toISOString(),
      activated_at: null,
      recall_initiated_at: null,
      recall_notice_deadline_at: null,
      return_custodian_ref: null,
      return_instruction_at: null,
      settled_at: null,
      defaulted_at: null,
    };
    mockLoans = [loan, ...mockLoans];
    return HttpResponse.json({ loan_id: loanId, state: 'pending' }, { status: 201 });
  }),

  http.get('/api/loans', async () => {
    await delay(20);
    const response: LoanListResponse = { loans: mockLoans };
    return HttpResponse.json(response);
  }),

  http.get('/api/loans/:loan_id', async ({ params }) => {
    await delay(20);
    const loan = mockLoans.find((row) => row.loan_id === params.loan_id);
    if (!loan) return mockError('not_found', 'Loan not found', 404);
    return HttpResponse.json(loan);
  }),
];

File: backend/app/schemas/loans.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field, model_validator


LoanState = Literal["pending", "active", "margin_call", "recall_initiated", "settled", "defaulted"]


class ApprovedBorrowerRequest(BaseModel):
    borrower_id: uuid.UUID


class ApprovedBorrowerResponse(BaseModel):
    borrower_id: uuid.UUID
    name: str
    jurisdiction: str
    status: str
    approved_at: datetime


class ApprovedBorrowerListResponse(BaseModel):
    borrowers: list[ApprovedBorrowerResponse]


class LoanBookingRequest(BaseModel):
    connection_id: uuid.UUID
    borrower_id: uuid.UUID
    asset_type: str = Field(min_length=1)
    quantity: Decimal = Field(gt=0)
    rate_bps: int = Field(ge=0)
    term_type: Literal["open", "fixed"]
    maturity_date: date | None = None
    collateral_type: str = Field(min_length=1)
    collateral_quantity: Decimal = Field(gt=0)
    collateral_value_usd: Decimal = Field(gt=0)

    @model_validator(mode="after")
    def fixed_term_requires_maturity(self) -> "LoanBookingRequest":
        if self.term_type == "fixed" and self.maturity_date is None:
            raise ValueError("fixed-term loans require maturity_date")
        if self.term_type == "open" and self.maturity_date is not None:
            raise ValueError("open-term loans must not include maturity_date")
        return self


class LoanCreateResponse(BaseModel):
    loan_id: uuid.UUID
    state: LoanState


class LoanResponse(BaseModel):
    loan_id: uuid.UUID
    connection_id: uuid.UUID
    agreement_id: uuid.UUID
    borrower_id: uuid.UUID
    borrower_name: str
    asset_type: str
    quantity: str
    rate_bps: int
    term_type: str
    maturity_date: date | None
    day_count_basis: str
    collateral_type: str
    collateral_quantity: str
    collateral_value_usd: str
    current_ltv_pct: str | None
    ltv_as_of: datetime | None
    state: LoanState
    booked_at: datetime
    activated_at: datetime | None
    recall_initiated_at: datetime | None
    recall_notice_deadline_at: datetime | None
    return_custodian_ref: str | None
    return_instruction_at: datetime | None
    settled_at: datetime | None
    defaulted_at: datetime | None


class LoanListResponse(BaseModel):
    loans: list[LoanResponse]


class CollateralSubstitutionRequest(BaseModel):
    collateral_type: str = Field(min_length=1)
    collateral_quantity: Decimal = Field(gt=0)
    collateral_value_usd: Decimal = Field(gt=0)
